### PR TITLE
Refactoring of handleDriverContainer code

### PR DESCRIPTION
**Pull request by yevgeny-shnaidman**
This PR refactors the code of handleDrivercontainer code, and as a result also refactors the moduleReconciler loop, garbage collector and status module updator The main points:
1) change the name of the module loader DS name creation. Now the name consists of
   module.Name + hash(kernel version). This way we don't actually need a map of DS per
   kernel version, we can allow CreateOrPatch function to do its work, since now DS name
   is deterministic
2) as a result of 1), the API of handleDriverContainer, garbageCollector, statusUpdater
   and reconcile loop has been change
3) DaemonSet interface now returns just a list of DS for module 4) the differintiation between device-plugin and module-loader DS
   is now based on the role label
5) updating unit test

Part of [263](https://github.com/kubernetes-sigs/kernel-module-management/issues/263)

### CODE REVIEW

**Comment by qbarrand** [comment]
@yevgeny-shnaidman the PR decreases the coverage, could you please fix that?

File: controllers/module_reconciler.go [hunk 1]
```diff
@@ -340,21 +345,18 @@ func (mrh *moduleReconcilerHelper) handleSigning(ctx context.Context, mld *api.M
 }
 
 func (mrh *moduleReconcilerHelper) handleDriverContainer(ctx context.Context,
-	mld *api.ModuleLoaderData,
-	dsByKernelVersion map[string]*appsv1.DaemonSet) error {
+	mld *api.ModuleLoaderData) error {
+
+	hashValue, err := hashstructure.Hash(hashData{kernelVersion: mld.KernelVersion}, nil)
```
Comment (qbarrand) [inline controllers/module_reconciler.go:350]: If we just hash one string, I don't see the benefit of using hashstructure. It is probably much more expensive than [fnv](https://pkg.go.dev/hash/fnv).

**Comment by qbarrand** [comment]
What is the benefit of dropping the map and relying on a hash? I originally went with that map to eliminate the risk of any collision.

File: controllers/module_reconciler.go (hunk 1 above)
Comment (yevgeny-shnaidman) [reply controllers/module_reconciler.go:350]: i want to use hashstructure, since later on we will add a module Version to the hash

**Comment by yevgeny-shnaidman** [comment]
> What is the benefit of dropping the map and relying on a hash? I originally went with that map to eliminate the risk of any collision.

hash is safe enough, especially since we are talking about DaemonSets  per module per namespace. The benefit is simplifying the code, eliminating double code, since CreateOrPatch check for DS existance and knows to take the existent DS and patch it. In addition, when we will work with ModuleVersion, we don't need to maintain map per kernel version per module version - we can just code module version as part of DS name and the same code will work

File: internal/daemonset/daemonset.go [hunk 2]
```diff
@@ -52,12 +53,13 @@ func NewCreator(client client.Client, kernelLabel string, scheme *runtime.Scheme
...
-func (dc *daemonSetGenerator) GarbageCollect(ctx context.Context, existingDS map[string]*appsv1.DaemonSet, validKernels sets.Set[string]) ([]string, error) {
+func (dc *daemonSetGenerator) GarbageCollect(ctx context.Context, existingDS []appsv1.DaemonSet, validKernels sets.Set[string]) ([]string, error) {
 	deleted := make([]string, 0)
 
-	for kernelVersion, ds := range existingDS {
-		if !dc.isDevicePluginDaemonSet(ds) && !validKernels.Has(kernelVersion) {
-			if err := dc.client.Delete(ctx, ds); err != nil {
+	for _, ds := range existingDS {
+		dsKernelVersion := ds.Labels[dc.kernelLabel]
+		if !IsDevicePluginDS(&ds) && !validKernels.Has(dsKernelVersion) {
```
Comment (ybettan) [inline internal/daemonset/daemonset.go:61]:
```suggestion
		
		if dsKernelVersion := ds.Labels[dc.kernelLabel]; !IsDevicePluginDS(&ds) && !validKernels.Has(dsKernelVersion) {
```

This variable is only scoped to the `if` block

**Comment by qbarrand** [comment]
> In addition, when we will work with ModuleVersion, we don't need to maintain map per kernel version per module version

What is the cost of that? I think a struct with both the kernel and the module version can be a map key.

File: internal/daemonset/daemonset.go (hunk 2 above)
Comment (yshnaidman) [reply internal/daemonset/daemonset.go]: fixed

File: controllers/module_reconciler.go (hunk 1 above)
Comment (qbarrand) [reply controllers/module_reconciler.go:350]: Turns out hashstructure does use fnv.

File: controllers/module_reconciler.go [hunk 3]
```diff
@@ -340,21 +345,18 @@ func (mrh *moduleReconcilerHelper) handleSigning(ctx context.Context, mld *api.M
...
+	if err != nil {
+		return fmt.Errorf("failed to hash kernel version %s for module loader daemonset name: %v", mld.KernelVersion, err)
+	}
+	name := mld.Name + "-" + fmt.Sprintf("%d", hashValue)
```
Comment (qbarrand) [inline controllers/module_reconciler.go:354]: How about we use hex to same some characters? Also, let's do all the formatting in `fmt.Sprintf`:
```suggestion
	name := fmt.Sprintf("%s-%x", mld.Name, hashValue)
```

File: controllers/module_reconciler.go [hunk 4]
```diff
@@ -31,6 +31,7 @@ import (
 	"github.com/kubernetes-sigs/kernel-module-management/internal/sign"
 	"github.com/kubernetes-sigs/kernel-module-management/internal/statusupdater"
 	"github.com/kubernetes-sigs/kernel-module-management/internal/utils"
+	"github.com/mitchellh/hashstructure"
```
Comment (qbarrand) [inline controllers/module_reconciler.go:34]: Let's use v2.

Comment (yevgeny-shnaidman) [reply controllers/module_reconciler.go]: fixed

File: controllers/module_reconciler.go (hunk 3 above)
Comment (yevgeny-shnaidman) [reply controllers/module_reconciler.go]: fixed

File: controllers/module_reconciler.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply controllers/module_reconciler.go]: fixed

File: internal/statusupdater/statusupdater_test.go [hunk 5]
```diff
@@ -256,25 +255,26 @@ var _ = Describe("preflight status updates", func() {
...
+	return ds
 }
 
-func prepareDsByKernel(dsConfigs []daemonSetConfig) map[string]*appsv1.DaemonSet {
-	dsMap := make(map[string]*appsv1.DaemonSet)
-	for i, dsConfig := range dsConfigs {
-		kernelVersion, ds := getDaemonSet(i, dsConfig)
-		dsMap[kernelVersion] = ds
+func prepareDsByKernel(dsConfigs []daemonSetConfig) []appsv1.DaemonSet {
+	dsSlice := []appsv1.DaemonSet{}
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply internal/statusupdater/statusupdater_test.go]: fixed

File: controllers/module_reconciler_test.go [hunk 6]
```diff
@@ -555,20 +556,18 @@ var _ = Describe("ModuleReconciler_handleDriverContainer", func() {
...
 			KernelVersion: "kernelVersion1",
 		}
-		existingDS := map[string]*appsv1.DaemonSet{
-			"kernelVersion2": &appsv1.DaemonSet{}, "kernelVersion3": &appsv1.DaemonSet{},
-		}
+		hashValue, err := hashstructure.Hash(hashData{kernelVersion: mld.KernelVersion}, nil)
+		Expect(err).NotTo(HaveOccurred())
 		newDS := &appsv1.DaemonSet{
-			ObjectMeta: metav1.ObjectMeta{Namespace: mld.Namespace},
+			ObjectMeta: metav1.ObjectMeta{Namespace: mld.Namespace, Name: mld.Name + "-" + fmt.Sprintf("%d", hashValue)},
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply controllers/module_reconciler_test.go]: fixed

**Pull request merged**